File: behavioral/state/order_processing/typescript/order.ts
import { IOrderState, NewOrderState } from "./state";

/**
 * The Context class representing an order.
 */
export class Order {
  public items: string[] = [];
  private state: IOrderState;
  public readonly orderId: string;

  constructor(orderId: string) {
    this.orderId = orderId;
    // Default state is NewOrderState
    this.state = new NewOrderState(this);
    console.log(
      `Order ${this.orderId} created in state: ${this.state.toString()}`
    );
  }

  /**
   * Allows changing the state.
   */
  public setState(state: IOrderState): void {
    console.log(
      `Order ${
        this.orderId
      }: Transitioning from ${this.state.toString()} to ${state.toString()}`
    );
    this.state = state;
  }

  /**
   * Returns the current state.
   */
  public getState(): IOrderState {
    return this.state;
  }

  public addItem(item: string): void {
    this.state.addItem(item);
  }

  public processPayment(): void {
    this.state.processPayment();
  }

  public ship(): void {
    this.state.ship();
  }

  public deliver(): void {
    this.state.deliver();
  }

  public cancel(): void {
    this.state.cancel();
  }

  public toString(): string {
    return `Order [ID: ${
      this.orderId
    }, State: ${this.state.toString()}, Items: [${this.items.join(", ")}]]`;
  }
}

File: behavioral/state/order_processing/python/order.py
from __future__ import annotations
from state import OrderState, NewOrderState
from typing import List

class Order:
    """The Context class representing an order."""
    def __init__(self, order_id: str):
        self.order_id = order_id
        self.items: List[str] = []
        # Default state is NewOrderState
        self._state: OrderState = NewOrderState(self)
        print(f"Order {self.order_id} created in state: {self._state}")

    def set_state(self, state: OrderState):
        """Allows changing the state."""
        print(f"Order {self.order_id}: Transitioning from {self._state} to {state}")
        self._state = state

    def get_state(self) -> OrderState:
        """Returns the current state."""
        return self._state

    def add_item(self, item: str):
        """Delegate action to the current state."""
        self._state.add_item(item)

    def process_payment(self):
        """Delegate action to the current state."""
        self._state.process_payment()

    def ship(self):
        """Delegate action to the current state."""
        self._state.ship()

    def deliver(self):
        """Delegate action to the current state."""
        self._state.deliver()

    def cancel(self):
        """Delegate action to the current state."""
        self._state.cancel()

    def __str__(self) -> str:
        return f"Order [ID: {self.order_id}, State: {self._state}, Items: {self.items}]"